### PR TITLE
✨ feat: add selectable types to Calendar

### PR DESCRIPTION
**Pull request by hbatalhaStch**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link

close https://github.com/ant-design/ant-design/issues/40407

### 💡 Background and solution

This PR follows [this one](https://github.com/ant-design/ant-design/pull/42112).

When I need to go to a specific date in a year and month different from the current selected date, it will require me to select the date 3 times just to get there. First I have to select the year, then select the month and only then select the day which is exhaustive and unnecessary IMO as it gives a really bad UX.
The issue is that the onSelect callback is called whenever we change the month or year. I propose an api that let you choose when you want the onSelect callback to be called, if it is when we change the year, when we change the month or only when change click on a day. We can also combine them.

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |      Make calendar selectable by `date`, `month` and `year`     |
| 🇨🇳 Chinese |      使日历可以按`date`、`month`和`year`进行选择      |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

File: components/calendar/index.en-US.md [hunk 1]
```diff
@@ -55,6 +55,7 @@ When data is in the form of dates, such as schedules, timetables, prices calenda
 | value | The current selected date | [dayjs](https://day.js.org/) | - |  |
 | onChange | Callback for when date changes | function(date: Dayjs) | - |  |
 | onPanelChange | Callback for when panel changes | function(date: Dayjs, mode: string) | - |  |
+| selectable | Make calendar selectable by `date`, `month` and `year` | ('date' \| 'year' \| 'month')[] | - |  |
```
Comment (yoyo837) [inline components/calendar/index.en-US.md:58]:
```suggestion
| selectable | Make calendar selectable by `date`, `month` and `year` | ('date' \| 'year' \| 'month')[] | - | 5.6.0 |
```

File: components/calendar/index.zh-CN.md [hunk 2]
```diff
@@ -60,6 +60,7 @@ coverDark: https://mdn.alipayobjects.com/huamei_7uahnr/afts/img/A*-p-wQLik200AAA
 | value | 展示日期 | [dayjs](https://day.js.org/) | - |  |
 | onChange | 日期变化回调 | function(date: Dayjs) | - |  |
 | onPanelChange | 日期面板变化回调 | function(date: Dayjs, mode: string) | - |  |
+| selectable | 使日历可以按 `date`、`month` 和 `year` 进行选择 | ('date' \| 'year' \| 'month')[] | - |  |
```
Comment (yoyo837) [inline components/calendar/index.zh-CN.md:63]:
```suggestion
| selectable | 使日历可以按 `date`、`month` 和 `year` 进行选择 | ('date' \| 'year' \| 'month')[] | - | 5.6.0 |
```

**Comment by afc163** [comment]
Wait for @zombieJ 's decision.

**Comment by hbatalhaStch** [comment]
> Need to add docs for `triggerType` field.

Where should I add it in the docs?

**Comment by yoyo837** [comment]
> > Need to add docs for `triggerType` field.
> 
> Where should I add it in the docs?

For example, you can refer to the docs at this location:
https://ant.design/components/table#api

image

**Comment by hbatalhaStch** [comment]
> For example, you can refer to the docs at this location:

Like this?:
```md
### Calendar

| Property | Description | Type | Default | Version |
| --- | --- | --- | --- | --- |
| dateCellRender | Customize the display of the date cell, the returned content will be appended to the cell | function(date: Dayjs): ReactNode | - |  |
| dateFullCellRender | Customize the display of the date cell, the returned content will override the cell | function(date: Dayjs): ReactNode | - |  |
| defaultValue | The date selected by default | [dayjs](https://day.js.org/) | - |  |
| disabledDate | Function that specifies the dates that cannot be selected, `currentDate` is same dayjs object as `value` prop which you shouldn't mutate it](https://github.com/ant-design/ant-design/issues/30987) | (currentDate: Dayjs) => boolean | - |  |
| fullscreen | Whether to display in full-screen | boolean | true |  |
| headerRender | Config of headerRender. You can ref calendar headerRender [config](#headerRender) | object  | - |  |
| locale | The calendar's locale | object | [(default)](https://github.com/ant-design/ant-design/blob/master/components/date-picker/locale/example.json) |  |
| mode | The display mode of the calendar | `month` \| `year` | `month` |  |
| monthCellRender | Customize the display of the month cell, the returned content will be appended to the cell | function(date: Dayjs): ReactNode | - |  |
| monthFullCellRender | Customize the display of the month cell, the returned content will override the cell | function(date: Dayjs): ReactNode | - |  |
| validRange | To set valid range | \[[dayjs](https://day.js.org/), [dayjs](https://day.js.org/)] | - |  |
| value | The current selected date | [dayjs](https://day.js.org/) | - |  |
| onChange | Callback for when date changes | function(date: Dayjs) | - |  |
| onPanelChange | Callback for when panel changes | function(date: Dayjs, mode: string) | - |  |
| selectable | Make calendar selectable by `date`, `month` and `year` | ('date' \| 'year' \| 'month')[] | - |  |
| onSelect | Callback for when a date is selected | function(date: Dayjs） | - |  |

### headerRender

Properties for headerRender.

| Property | Description | Type | Default |
| --- | --- | --- | --- |
| value | Current date | [dayjs](https://day.js.org/) | - |
| type | Calendar | `month` \| `year` | - |
| onChange |  Callback for when date changes | function(date: Dayjs, selectInfo?: { triggerType?: CalendarSelectable }) | - |
| onTypeChange | Callback for when mode changes | function(type: `month` \| `year`) | - |

```

refer to `headerRender`

**Comment by hbatalhaStch** [comment]
Close in favor of https://github.com/ant-design/ant-design/pull/42432

**Pull request closed without merging**